### PR TITLE
Change `validateTextLinks` to only get URL in markdown links

### PR DESCRIPTION
**Pull request by GuillaumeRx**
This updates `validateTextLinks` to only search for URLs in Mardown link formatting.

- Update Regex
- Update test
- Use `String.matchAll()`

### CODE REVIEW

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

File: packages/snaps-ui/src/validation.ts [hunk 1]
```diff
@@ -55,13 +55,13 @@ export function assertLinksAreSafe(
           `Protocol must be one of: ${ALLOWED_PROTOCOLS.join(', ')}.`,
         );
 
-        const hostname =
+        const targetUrl =
           url.protocol === 'mailto:'
             ? url.pathname.split('@')[1]
-            : url.hostname;
+            : url.toString();
```
Comment (Mrtenz) [inline packages/snaps-ui/src/validation.ts:61]: Please add a test to make sure that the behaviour is now correct.

**Review by Mrtenz — CHANGES_REQUESTED** [review]
Please add a PR description as well.

**Comment by GuillaumeRx** [comment]
@metamaskbot publish-preview

File: packages/snaps-utils/src/ui.ts [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
+import { assert, AssertionError } from '@metamask/utils';
+
+const MARKDOWN_LINK_REGEX = /!?\[(?<name>[^\]]*)\]\((?<url>[^)]+)\)/giu;
 
-const LINK_REGEX = /(?<protocol>[a-z]+:\/?\/?)(?<host>\S+?(?:\.[a-z]+)+)/giu;
 const ALLOWED_PROTOCOLS = ['https:', 'mailto:'];
 
 /**
- * Search for links in a sting and check them against the phishing list.
+ * Searches for markdown links in a sting and checks them against the phishing list.
```
Comment (FrederikBolding) [inline packages/snaps-utils/src/ui.ts:10]:
```suggestion
 * Searches for markdown links in a string and checks them against the phishing list.
```

File: packages/snaps-utils/src/ui.ts [hunk 3]
```diff
@@ -17,35 +18,39 @@ export function validateTextLinks(
...
-        assert(
-          ALLOWED_PROTOCOLS.includes(url.protocol),
-          `Protocol must be one of: ${ALLOWED_PROTOCOLS.join(', ')}.`,
-        );
+  const matches = text.matchAll(MARKDOWN_LINK_REGEX);
+  if (matches) {
+    for (const { groups } of matches) {
+      const link = groups?.url;
+
+      if (link) {
```
Comment (FrederikBolding) [inline packages/snaps-utils/src/ui.ts:26]: Could we rewrite to reduce the nesting? It's a bit difficult to read now.

Can probably be done with 
```js
if (!link) {
  continue;
}
```

or a filter on the `matches`?

File: packages/snaps-utils/src/ui.ts [hunk 4]
```diff
@@ -17,9 +18,15 @@ export function validateTextLinks(
   text: string,
   isOnPhishingList: (url: string) => boolean,
 ) {
-  const links = text.match(LINK_REGEX);
-  if (links) {
-    links.forEach((link) => {
+  const matches = String.prototype.matchAll.call(text, MARKDOWN_LINK_REGEX);
+  if (matches) {
```
Comment (FrederikBolding) [inline packages/snaps-utils/src/ui.ts:22]: Maybe we can unnest here too?

```js
if (!matches) {
  return;
}
```

What do you think?

File: packages/snaps-utils/src/ui.ts [hunk 5]
```diff
@@ -1,12 +1,13 @@
 import type { Component } from '@metamask/snaps-sdk';
 import { NodeType } from '@metamask/snaps-sdk';
-import { assert, AssertionError, hasProperty } from '@metamask/utils';
+import { assert, AssertionError } from '@metamask/utils';
+
+const MARKDOWN_LINK_REGEX = /!?\[(?<name>[^\]]*)\]\((?<url>[^)]+)\)/giu;
```
Comment (FrederikBolding) [inline packages/snaps-utils/src/ui.ts:5]: Why do we need the `!?` in this regex?

Comment (GuillaumeRx) [reply packages/snaps-utils/src/ui.ts]: Yeah, it's not needed. Good catch

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**